### PR TITLE
NDRS-650: Send finality signatures to the event stream.

### PR DESCRIPTION
**Pull request by goral09**
https://casperlabs.atlassian.net/browse/NDRS-650

I suggest reviewing commit-by-commit.

### CODE REVIEW

File: node/src/types/block.rs [hunk 1]
```diff
@@ -1089,6 +1090,43 @@ pub(crate) mod json_compatibility {
     }
 }
 
+/// Finality signature that can be gossiped between nodes or sent to clients.
```
Comment (afck) [inline node/src/types/block.rs:1117]: Maybe:
`A validator's signature of a block, to confirm it is finalized. Clients and joining nodes should wait until the signers' combined weight exceeds their fault tolerance threshold before accepting the block as finalized.`

File: node/src/effect.rs [hunk 2]
```diff
@@ -254,11 +254,20 @@ pub trait EffectOptionExt {
     ///
     /// The function `f_some` is used to translate the returned value from an effect into an event,
     /// while the function `f_none` does the same for a returned `None`.
-    fn option<U, F, G>(self, f_some: F, f_none: G) -> Effects<U>
+    fn fold<U, F, G>(self, f_some: F, f_none: G) -> Effects<U>
```
Comment (afck) [inline node/src/effect.rs:257]: Shouldn't it rather be `map_or_else`, like [the corresponding method on `Option`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or_else)?

Comment (afck) [reply node/src/effect.rs:257]: Also, we should probably switch the order of its arguments, so it's in line with the standard library.

File: node/src/effect.rs [hunk 3]
```diff
@@ -254,11 +254,20 @@ pub trait EffectOptionExt {
...
     where
         F: FnOnce(Self::Value) -> U + 'static + Send,
         G: FnOnce() -> U + 'static + Send,
         U: 'static;
+
+    /// Finalizes a future returning an `Option` into two different effects.
+    ///
+    /// The function `f` is used to translate the returned value from an effect into an event,
+    /// In the case of `None`, empty vector of effects is returned.
+    fn map_some<U, F>(self, f: F) -> Effects<U>
```
Comment (afck) [inline node/src/effect.rs:267]: (Awesome! That's what I was missing when I tried to create finality signatures only if we're a validator! :tada:)

File: node/src/effect/announcements.rs [hunk 4]
```diff
@@ -195,6 +198,9 @@ impl Display for LinearChainAnnouncement {
             LinearChainAnnouncement::BlockAdded { block_hash, .. } => {
                 write!(f, "block added {}", block_hash)
             }
+            LinearChainAnnouncement::NewFinalitySignature(fs) => {
+                write!(f, "new finsliaty signature {}", fs.block_hash)
```
Comment (afck) [inline node/src/effect/announcements.rs:202]: :thinking:
```suggestion
                write!(f, "new finality signature {}", fs.block_hash)
```

File: node/src/effect.rs [hunk 5]
```diff
@@ -936,6 +960,19 @@ impl<REv> EffectBuilder<REv> {
             .await
     }
 
+    /// The linear chain has stored new finality signature.
```
Comment (afck) [inline node/src/effect.rs:963]:
```suggestion
    /// The linear chain has stored a new finality signature.
```

**Review by afck — COMMENTED** [review]
Looks great! Just a few nit-picks.

File: node/src/effect.rs (hunk 2 above)
Comment (goral09) [reply node/src/effect.rs:257]: I think this ordering of arguments is better and the standard library should change it :)

Comment (goral09) [reply node/src/effect.rs]: Done in [`4e187e2` (#635)](https://github.com/CasperLabs/casper-node/pull/635/commits/4e187e27d16153b0f933af336f0abae1cdb7263c)

File: node/src/types/block.rs (hunk 1 above)
Comment (goral09) [reply node/src/types/block.rs]: Done in [`4e187e2` (#635)](https://github.com/CasperLabs/casper-node/pull/635/commits/4e187e27d16153b0f933af336f0abae1cdb7263c)

**Review by afck — APPROVED** [review]

**Pull request merged**